File: core/src/main/java/com/lomekwi/cave/project/Project.java
package com.lomekwi.cave.project;

import static com.lomekwi.cave.util.i18n.I18N.i18n;

import com.badlogic.gdx.Gdx;
import com.google.common.eventbus.EventBus;
import com.google.common.eventbus.Subscribe;
import com.lomekwi.cave.pipeline.audio.AudioFrameSink;
import com.lomekwi.cave.resource.Resource;
import com.lomekwi.cave.timeline.SegFactory;
import com.lomekwi.cave.timeline.Timeline;
import com.lomekwi.cave.timeline.Track;
import com.lomekwi.cave.timeline.UndoManager;
import com.lomekwi.cave.timeline.playback.Playhead;
import com.lomekwi.cave.app.App;

import com.google.common.collect.ArrayListMultimap;
import com.google.common.collect.Multimap;

import java.io.File;
import java.io.IOException;
import java.io.ObjectInputStream;
import java.io.Serial;
import java.io.Serializable;
import java.nio.file.Path;
import java.util.UUID;
import java.util.concurrent.Future;

public class Project implements Serializable, AutoCloseable {
    @Serial
    private static final long serialVersionUID = 1L;
    protected transient Path savePath;
    public final Timeline timeline;
    public transient Playhead playhead;
    public final Multimap<File, Resource> resources = ArrayListMultimap.create();
    public final SegFactory segFactory = new SegFactory(this);
    public transient EventBus projEventBus;
    public transient UndoManager undoManager;
    public String name;
    public final UUID uuid = UUID.randomUUID();
    public long savedVersion = 0;
    public transient long currentVersion = 0;

    private transient boolean isActive = false;

    protected Project() {
        App.appEventBus.register(this);
        name = i18n("未命名");
        projEventBus = new EventBus(uuid.toString());
        undoManager = new UndoManager(this);
        projEventBus.register(new AudioFrameSink());
        projEventBus.register(this);
        timeline = new Timeline(this);
        playhead = new Playhead(projEventBus);
    }

    public void update() {
        if (!isActive) {
            return;
        }

        for (Track track : timeline.getTracks()) {
            if (track.getWorker().getFuture() == null || track.getWorker().getFuture().isDone()) {
                Future<?> future = App.workerExecutor.submit(track.getWorker());
                track.getWorker().setFuture(future);
            }
        }
    }

    @Subscribe
    public void onProjectFronted(ProjectFrontedEvent event) {
        if (!isActive) {
            isActive = true;
            Gdx.app.log("Project", "项目 [" + name + "] 激活，开始轨道循环");
            update();
        }
    }

    @Subscribe
    public void onProjectBackgrounded(ProjectBackgroundedEvent event) {
        if (isActive) {
            isActive = false;
            Gdx.app.log("Project", "项目 [" + name + "] 后台化，停止轨道循环");
            stopTrackLoops();
        }
    }

    private void stopTrackLoops() {
        for (Track track : timeline.getTracks()) {
            Future<?> future = track.getWorker().getFuture();
            if (future != null) {
                future.cancel(true);
                track.getWorker().setFuture(null);
            }
        }
    }


    public void close() {
        App.appEventBus.unregister(this);
        isActive = false;
        stopTrackLoops();
        resources.values().forEach(resource -> {
            try {
                resource.close();
            } catch (Exception e) {
                throw new RuntimeException(e);
            }
        });
    }

    @Serial
    private void readObject(ObjectInputStream in) throws IOException, ClassNotFoundException {
        in.defaultReadObject();
        currentVersion = savedVersion;
        segFactory.setProject(this);
        App.appEventBus.register(this);
        projEventBus = new EventBus(uuid.toString());
        projEventBus.register(new AudioFrameSink());
        projEventBus.register(this);
        playhead = new Playhead(projEventBus);
        undoManager = new UndoManager(this);
        isActive = false;
    }

    public Path getSavePath() {
        return savePath;
    }

    @Serial
    private void writeObject(java.io.ObjectOutputStream out) throws IOException {
        savedVersion = currentVersion;
        out.defaultWriteObject();
    }

    public boolean isDirty() {
        return currentVersion != savedVersion;
    }
}

File: core/src/main/scala/com/lomekwi/cave/project/Project.scala
package com.lomekwi.cave.project

import com.lomekwi.cave.util.i18n.I18N.i18n

import com.badlogic.gdx.Gdx
import com.google.common.eventbus.EventBus
import com.google.common.eventbus.Subscribe
import com.lomekwi.cave.pipeline.audio.AudioFrameSink
import com.lomekwi.cave.resource.Resource
import com.lomekwi.cave.timeline.SourceFactory
import com.lomekwi.cave.timeline.Timeline
import com.lomekwi.cave.timeline.UndoManager
import com.lomekwi.cave.timeline.playback.Playhead
import com.lomekwi.cave.app.App

import com.google.common.collect.ArrayListMultimap
import com.google.common.collect.Multimap

import java.io.File
import java.io.ObjectInputStream
import java.io.Serializable
import java.nio.file.Path
import java.util.UUID

import scala.compiletime.uninitialized
import scala.jdk.CollectionConverters.*

@SerialVersionUID(2L)
class Project protected[project] extends Serializable with AutoCloseable {
  @transient protected[project] var savePath: Path = uninitialized
  var timeline: Timeline = uninitialized
  @transient var playhead: Playhead = uninitialized
  final val resources: Multimap[File, Resource] = ArrayListMultimap.create[File, Resource]()
  final val sourceFactory: SourceFactory = new SourceFactory(this)
  @transient var projEventBus: EventBus = uninitialized
  @transient var undoManager: UndoManager = uninitialized
  var name: String = uninitialized
  private final val uuid: UUID = UUID.randomUUID()
  var savedVersion: Long = 0
  @transient var currentVersion: Long = 0

  @transient private var isActive: Boolean = false

  App.appEventBus.register(this)
  name = i18n("未命名")
  projEventBus = new EventBus(uuid.toString)
  undoManager = new UndoManager(this)
  projEventBus.register(new AudioFrameSink(this))
  projEventBus.register(this)
  timeline = new Timeline(this)
  playhead = new Playhead(projEventBus)

  def update(): Unit = {
    if (isActive) {
      for (track <- timeline.getTracks.asScala) {
        if (track.getWorker.getFuture == null || track.getWorker.getFuture.isDone) {
          val future = App.workerExecutor.submit(track.getWorker)
          track.getWorker.setFuture(future)
        }
      }
    }
  }

  @Subscribe
  def onProjectFronted(event: ProjectFrontedEvent): Unit = {
    if (!isActive) {
      isActive = true
      Gdx.app.log("Project", "项目 [" + name + "] 激活，开始轨道循环")
      update()
    }
  }

  @Subscribe
  def onProjectBackgrounded(event: ProjectBackgroundedEvent): Unit = {
    if (isActive) {
      isActive = false
      Gdx.app.log("Project", "项目 [" + name + "] 后台化，停止轨道循环")
      stopTrackLoops()
    }
  }

  private def stopTrackLoops(): Unit = {
    for (track <- timeline.getTracks.asScala) {
      val future = track.getWorker.getFuture
      if (future != null) {
        future.cancel(true)
        track.getWorker.setFuture(null)
      }
    }
  }


  override def close(): Unit = {
    projEventBus.post(ProjectBackgroundedEvent)
    App.appEventBus.unregister(this)
    isActive = false
    stopTrackLoops()
    resources.values().forEach((resource: Resource) => {
      try {
        resource.close()
      } catch {
        case e: Exception =>
          throw new RuntimeException(e)
      }
    })
  }

  private def readObject(in: ObjectInputStream): Unit = {
    in.defaultReadObject()
    currentVersion = savedVersion
    sourceFactory.setProject(this)
    App.appEventBus.register(this)
    projEventBus = new EventBus(uuid.toString)
    projEventBus.register(new AudioFrameSink(this))
    projEventBus.register(this)
    playhead = new Playhead(projEventBus)
    undoManager = new UndoManager(this)
    isActive = false
  }

  def getSavePath: Path = {
    savePath
  }

  private def writeObject(out: java.io.ObjectOutputStream): Unit = {
    savedVersion = currentVersion
    out.defaultWriteObject()
  }

  def isDirty: Boolean = {
    currentVersion != savedVersion
  }
}
